pkg/gohai/network/network.go

```diff
@@ -6,29 +6,292 @@
+
+// Interface holds information about a specific interface
+type Interface struct {
+	// Name is the name of the interface
+	Name string `json:"name"`
+	// IPv4Network is the ipv4 address for the host
+	IPv4Network utils.Value[string] `json:"ipv4-network"`
+	// IPv6Network is the ipv6 address for the host
+	IPv6Network utils.Value[string] `json:"ipv6-network"`
+	// MacAddress is the mac address for the host
+	MacAddress utils.Value[string] `json:"macaddress"`
+	// IPv4 is the list of IPv4 addresses for the interface
+	IPv4 []string `json:"ipv4"`
+	// IPv4 is the list of IPv6 addresses for the interface
+	IPv6 []string `json:"ipv6"`
+}
+
+// Info holds network metadata about the host
+type Info struct {
+	// Interfaces is the list of interfaces which are up
+	Interfaces []Interface `json:"interfaces"`
+	// MacAddress is a mac address of the host
+	MacAddress string `json:"macaddress"`
+	// IPAddress is an IPv4 address of the host
+	IPAddress string `json:"ipaddress"`
+	// IPAddressV6 is an IPv6 address of the host
+	IPAddressV6 utils.Value[string] `json:"ipaddressv6"`
+}
+
+// CollectInfo collects the network information.
+func CollectInfo() (*Info, error) {
+	info := &Info{}
+	err := fillNetworkInfo(info)
+	if err != nil {
+		return info, err
+	}
+
+	interfaces, err := getMultiNetworkInfo()
+	if err == nil && len(interfaces) > 0 {
+		info.Interfaces = interfaces
+	}
+	return info, err
+}
+
+func ifacesToJSON(ifaces []Interface) (interface{}, []string) {
+	ret := make([]interface{}, len(ifaces))
+	warnings := []string{}
+
+	for idx, iface := range ifaces {
+		ifaceJSON := map[string]interface{}{
+			"name": iface.Name,
+			"ipv4": iface.IPv4,
+			"ipv6": iface.IPv6,
+		}
+
+		values := map[string]utils.Value[string]{
+			"ipv4-network": iface.IPv4Network,
+			"ipv6-network": iface.IPv6Network,
+			"macaddress":   iface.MacAddress,
+		}
+		for key, value := range values {
+			if val, err := value.Value(); err == nil {
+				ifaceJSON[key] = val
+			} else {
+				warnings = append(warnings, fmt.Sprintf("%s: %s: %v", iface.Name, key, err))
+			}
+		}
+
+		ret[idx] = ifaceJSON
+	}
+
+	return ret, warnings
+}
+
+// AsJSON returns an interface which can be marshalled to a JSON and contains the value of non-errored fields.
+func (netInfo *Info) AsJSON() (interface{}, []string, error) {
+	interfaces, warnings := ifacesToJSON(netInfo.Interfaces)
+	ret := map[string]interface{}{
+		"macaddress": netInfo.MacAddress,
+		"ipaddress":  netInfo.IPAddress,
+		"interfaces": interfaces,
+	}
+
+	if ipv6, err := netInfo.IPAddressV6.Value(); err == nil {
+		ret["ipaddressv6"] = ipv6
+	} else {
+		warnings = append(warnings, err.Error())
+	}
+
+	return ret, warnings, nil
+}
+
+func getMultiNetworkInfo() ([]Interface, error) {
+	multiNetworkInfo := []Interface{}
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return multiNetworkInfo, err
+	}
+
+	for _, iface := range ifaces {
+		defaultAddrErr := fmt.Errorf("%s: %w", iface.Name, ErrAddressNotFound)
```

Comment:
Could we move this above the `for` sow we only do this once ?

Reply:
If you mean creating the default error, it uses the interface `Name` so I don't think we can

Reply:
Indeed, misread this part :+1: .